<launch>
  <!-- General arguments -->
  <arg name="pointcloud_topic"
       default="pointcloud_in"
       doc="ROS topic for pointcloud input"/>
  <arg name="param_file"
       default="$(find wavemap_2d_ros)/config/wavemap_2d_server.yaml"
       doc="Name of the file from which to read wavemap's server params."/>
  <arg name="rosbag_path"
       default=""
       doc="Replay the rosbag at the specified path. Can be left blank."/>
  <arg name="rosbag_skip_first_n_sec"
       default="0"
       doc="Skip the first n seconds of the rosbag."/>
  <arg name="debug_mode"
       default="false"
       doc="Run the mapper in a debugger (gdb)."/>
  <arg name="show_rviz"
       default="true"
       doc="Automatically launch Rviz."/>
  <!-- Wavemap params -->
  <arg name="data_structure_type"
       default="dense_grid"/>
  <arg name="measurement_model_type"
       default="beam_model"/>

  <!-- Play the rosbag if a path was specified -->
  <group unless="$(eval arg('rosbag_path') == '')">
    <param name="/use_sim_time" value="true"/>
    <node name="player" pkg="rosbag" type="play" output="screen"
          args="-r 1.0 -s $(arg rosbag_skip_first_n_sec) -q --clock $(arg rosbag_path)"/>
  </group>

  <!-- Run the wavemap server -->
  <group>
    <arg if="$(arg debug_mode)" name="launch_prefix"
         value="gdb -ex run --args"/>
    <arg unless="$(arg debug_mode)" name="launch_prefix" value=""/>

    <node name="wavemap_2d" pkg="wavemap_2d_ros"
          type="wavemap_2d_server_node"
          output="screen" args="-alsologtostderr"
          launch-prefix="$(arg launch_prefix)">
      <rosparam file="$(arg param_file)"/>
      <param name="pointcloud_topic_name" value="$(arg pointcloud_topic)"/>
      <param name="data_structure_type" value="$(arg data_structure_type)"/>
      <param name="measurement_model_type" value="$(arg measurement_model_type)"/>
    </node>
  </group>

  <!-- Show Rviz -->
  <group if="$(arg show_rviz)">
    <arg name="rviz_file"
         default="$(find wavemap_2d_ros)/config/wavemap_2d_server.rviz"/>
    <node pkg="rviz" type="rviz" name="$(anon rviz)" output="screen"
          args="-d $(arg rviz_file)"/>
  </group>
</launch>
